---
[v1] net: phy: add eee advertisement2 broken modes support

The PHY EEE broken mode handling currently only covers modes
advertised via MDIO_AN_EEE_ADV (IEEE 802.3-2018 Register 7.60),
which includes 100TX, 1000T, 10GT, 1000KX, 10GKX4, and 10GKR.

IEEE 802.3-2022 introduced a second advertisement register
MDIO_AN_EEE_ADV2 (Register 7.62) for 2.5GBASE-T and 5GBASE-T EEE
modes. This patch extends the existing broken mode support to also
cover these two new speeds by introducing the 'eee-broken-2_5gt'
and 'eee-broken-5gt' device tree properties.

Without this patch, the PHY with no 2.5G/5G EEE support could
still advertise and negotiate these modes, leading to link
failures or data transmission issues.
---
v1: Add EEE advertisement2 broken modes support.
---

From 77a8efdc5d1a1d4b42296cc3c95ffa47c10f420a Mon Sep 17 00:00:00 2001
From: Mason Chang <mason-cw.chang@mediatek.com>
Date: Thu, 27 Aug 2026 14:30:51 +0800
Subject: [PATCH] net: phy: add eee advertisement2 broken modes support

This patch extends the existing broken mode support to also
cover these two new speeds by introducing the 'eee-broken-2_5gt'
and 'eee-broken-5gt' device tree properties.

Without this patch, the PHY with no 2.5G/5G EEE support could
still advertise and negotiate these modes, leading to link
failures or data transmission issues.

---
 drivers/net/phy/phy-c45.c  | 5 +++++
 drivers/net/phy/phy-core.c | 8 ++++++++
 include/linux/phy.h        | 4 ++++
 3 files changed, 17 insertions(+)

diff --git a/drivers/net/phy/phy-c45.c b/drivers/net/phy/phy-c45.c
index 5695935..08e24db 100644
--- a/drivers/net/phy/phy-c45.c
+++ b/drivers/net/phy/phy-c45.c
@@ -710,6 +710,11 @@ int genphy_c45_write_eee_adv(struct phy_device *phydev, unsigned long *adv)
 	if (linkmode_intersects(phydev->supported_eee, PHY_EEE_CAP2_FEATURES)) {
 		val = linkmode_to_mii_eee_cap2_t(adv);
 
+		/* In eee_broken_modes_adv2 are stored MDIO_AN_EEE_ADV2 specific raw
+		 * register values.
+		 */
+		val &= ~phydev->eee_broken_modes_adv2;
+
 		/* IEEE 802.3-2022 45.2.7.16 EEE advertisement 2
 		 * (Register 7.62)
 		 */
diff --git a/drivers/net/phy/phy-core.c b/drivers/net/phy/phy-core.c
index 1f98b6a..e0b7e5f 100644
--- a/drivers/net/phy/phy-core.c
+++ b/drivers/net/phy/phy-core.c
@@ -410,6 +410,14 @@ void of_set_phy_eee_broken(struct phy_device *phydev)
 		broken |= MDIO_EEE_10GKR;
 
 	phydev->eee_broken_modes = broken;
+
+	broken = 0;
+	if (of_property_read_bool(node, "eee-broken-2_5gt"))
+		broken |= MDIO_EEE_2_5GT;
+	if (of_property_read_bool(node, "eee-broken-5gt"))
+		broken |= MDIO_EEE_5GT;
+
+	phydev->eee_broken_modes_adv2 = broken;
 }
 
 /**
diff --git a/include/linux/phy.h b/include/linux/phy.h
index 1c3462a..716d60d 100644
--- a/include/linux/phy.h
+++ b/include/linux/phy.h
@@ -617,6 +617,8 @@ struct macsec_ops;
  * @lp_advertising: Current link partner advertised linkmodes
  * @host_interfaces: PHY interface modes supported by host
  * @eee_broken_modes: Energy efficient ethernet modes which should be prohibited
+ * @eee_broken_modes_adv2: Energy efficient ethernet modes from advertisment 2
+ *                         register which should be prohibited
  * @autoneg: Flag autoneg being used
  * @rate_matching: Current rate matching mode
  * @link: Current link state
@@ -738,6 +740,8 @@ struct phy_device {
 
 	/* Energy efficient ethernet modes which should be prohibited */
 	u32 eee_broken_modes;
+	u32 eee_broken_modes_adv2;
+
 	bool enable_tx_lpi;
 	struct eee_config eee_cfg;
 
-- 
2.45.2

